### ISSUE TITLE
2.6.1: flattener error (wrong answer)

### ISSUE THREAD

**Issue by matsc-at-sics-se** [MiniZinc/libminizinc#558]
`minizinc -a` on the following model finds 8 solutions, which is wrong. There are 6 solutions.
```
var 0..1: A;
var 0..1: B;
var 0..1: C;
constraint
    not((((B=0) -> (B=1)) -> (A=1)))
    ->
    not(((C=1) -> true));
```
`minizinc -a` on the following reports inconsistency, which is correct.
```
var 0..1: A;
var 0..1: B;
var 0..1: C;
constraint A=0 /\ B=1;
constraint
    not((((B=0) -> (B=1)) -> (A=1)))
    ->
    not(((C=1) -> true));
```
I found this by generating random models.

**Comment by krzku** [comment]
It seems that there is a problem in generation of *_imp constraints. If I commented all *_imp definitions in redefinitions.mzn the answers are correct. In this case, minizinc compiler generates _reif constraints only.

**Comment by Dekker1** [comment]
Hi Mats,

It appears that this issue was caused by an error in the context forwarding in the compiler. The clause in the LHS of the implication is converted to a formal, but it's context is not negated. I've pushed a fix for the problem and added some additional logic for the compilation of clauses that should avoid some more negations. Your model has been added as a regression test.

**Comment by informarte** [comment]
@Dekker1, going from MiniZinc 2.5.5 to 2.6.1, I noticed a change in the `tpp` models generated for Yuck: constraints like

```
constraint bool_clause_reif([X_INTRODUCED_335_,X_INTRODUCED_340_],[X_INTRODUCED_219_],X_INTRODUCED_341_):: defines_var(X_INTRODUCED_341_);
```

got replaced by

```
constraint bool_clause([X_INTRODUCED_335_,X_INTRODUCED_340_],[X_INTRODUCED_219_,X_INTRODUCED_341_]):: defines_var(X_INTRODUCED_341_);
```

with NO other changes.

I was able to show that the new constraints are produced by

```
predicate bool_clause_imp(array[int] of var bool: as, array[int] of var bool: bs, var bool: b) =
  clause(as,bs++[b]);
```

defined by `redefinitions-2.5.6.mzn`.

(I also observed this change for the `hrc` and `mario` instances.)

Can you confirm that these changes were caused by the bug you fixed?

**Comment by Dekker1** [comment]
Hi Michael,

These changes are not so much caused by a fixed bug, but rather an optimisation that was so far missing. As you might be aware, half-reifications (`r -> pred()`) can be used instead of reifications (`r <-> red`) when an expression is in positive context. In the MiniZinc library these are represented as `_imp` predicate.

Note that the half reification of a clause is still a clause, and thus I've added the redefinition to rewrite it as such. We recently actually had the discussion to remove this intermediate call since we couldn't come up with any kind of solver that would need to redefine it differently.

Note that the bug in this issue was in the context system itself, which is much older than the half-reification implementation. It was already used to reason about whether you can create new variables within an expression or not. This particular bug was caused by the optimisation mechanism of "pushing negations inwards" not being in sync with the contexts. (Needing multiple levels of nested clauses to trigger) Although you can never be sure that that are no more bugs, it should probably be noted that half-reifications have been included in the compiler since 2.3.0 and has been tested many times since then.

**Comment by informarte** [comment]
Hi Jip,  your definition of `bool_clause_imp` is problematic for local search because it replaces a functionally defined variable with a search variable. However, you explanation helped me to understand how to avoid this issue: I simply extended the Yuck library with an empty `redefinitions-2.5.6.mzn` file to restore the pre-2.6.0 behaviour. Thank you!
